aiohttp/cookiejar.py

```diff
@@ -256,8 +256,27 @@ def filter_cookies(
             and request_origin not in self._treat_as_secure_origin
         )
 
+        # d: domain
+        # p: path
+        d = hostname
+        p = request_url.path
+
+        cookies = list(self._cookies[(d, p)].values())
+
+        # shared cookie, it should have max of 1 entry
+        shared_cookie = self._cookies.get(("", "/"))
+        if shared_cookie:
+            shared_cookie_morsel = list(shared_cookie.values())[0]
+            cookies.append(shared_cookie_morsel)
+
+        # no path cookie, it should have max of 1 entry
+        no_path_cookie = self._cookies.get((d, "/"))
+        if no_path_cookie:
+            no_path_cookie_morsel = list(no_path_cookie.values())[0]
+            cookies.append(no_path_cookie_morsel)
+
         # Point 2: https://www.rfc-editor.org/rfc/rfc6265.html#section-5.4
-        for cookie in sorted(self, key=lambda c: len(c["path"])):
+        for cookie in sorted(cookies, key=lambda c: len(c["path"])):
```

Comment:
We should be able to remove the sorted() call at this point, because we'll have sorted the cookies with the loop we create above. I suspect there are other parts of the code we can remove too inside the loop, in order to make it more efficient (e.g. we shouldn't need to do the domain match, as we've already looked up the cookies by name).